containers/attributes/hooks.py

```diff
@@ -0,0 +1,82 @@
+import uuid
+import json
+import datetime
+import os
+import logging
+import sys
+from enum import Enum
+
+##### Set Up Logger ########
+
+AUDIT_LEVEL_NUM = os.getenv("AUDIT_LEVEL_NUM", 45)
+AUDIT_ENABLED = os.getenv("AUDIT_ENABLED", "false").lower() in ("yes", "true", "t", "1")
+
+logging.basicConfig(
+    stream=sys.stdout, level=os.getenv("SERVER_LOG_LEVEL", "CRITICAL").upper()
+)
+
+logging.addLevelName(AUDIT_LEVEL_NUM, "AUDIT")
+
+def audit(self, message, *args, **kws):
+    if self.isEnabledFor(AUDIT_LEVEL_NUM) and AUDIT_ENABLED:
+        self._log(AUDIT_LEVEL_NUM, message, args, **kws)
+
+logging.Logger.audit = audit
+logger = logging.getLogger(__package__)
+
+##### Custom Enums #########
+
+class CallType(Enum):
+    PRE = 1
+    POST = 2
+    ERR = 3
+
+class HttpMethod(Enum):
+    GET = 4
+    POST = 5
+    PUT = 6
+    PATCH = 7
+    DELETE = 8
+
+###### HOOKS ##########
+
+def run_pre_hooks(http_method, function_name, *args, **kwargs):
+    # STUB
+    pass
+
+def run_post_hooks(http_method, function_name, *args, **kwargs):
+    if http_method in [HttpMethod.POST, HttpMethod.PUT, HttpMethod.DELETE]:
+        _audit_log(CallType.POST, http_method, function_name, *args, **kwargs)
+    pass
+
+def run_err_hooks(http_method, function_name, err, *args, **kwargs):
```

Comment:
See above re: name things by what they do instead of where they go
```suggestion
def audit_failure_hook(http_method, function_name, err, *args, **kwargs):
```